lxd-agent/metrics.go

```diff
@@ -0,0 +1,375 @@
+		if !re.MatchString(entry.Name()) {
+			continue
+		}
+
+		cmdlinePath := filepath.Join("/proc", entry.Name(), "cmdline")
+
+		cmdline, err := ioutil.ReadFile(cmdlinePath)
+		if err != nil {
+			continue
+		}
+
+		if string(cmdline) == "" {
+			continue
+		}
+
+		pidCount++
+	}
+
+	return pidCount, nil
+}
+
+func getDiskMetrics(d *Daemon) (map[string]metrics.DiskMetrics, error) {
+	diskStats, err := ioutil.ReadFile("/proc/diskstats")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read /proc/diskstats: %w", err)
+	}
+
+	out := map[string]metrics.DiskMetrics{}
+	scanner := bufio.NewScanner(bytes.NewReader(diskStats))
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if line == "" {
+			continue
+		}
+
+		fields := strings.Fields(line)
+		stats := metrics.DiskMetrics{}
+
+		stats.ReadsCompleted, err = strconv.ParseUint(fields[3], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse %q: %w", fields[3], err)
+		}
+
+		sectorsRead, err := strconv.ParseUint(fields[5], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse %q: %w", fields[3], err)
+		}
+
+		stats.ReadBytes = sectorsRead * 512
+
+		stats.WritesCompleted, err = strconv.ParseUint(fields[7], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse %q: %w", fields[3], err)
+		}
+
+		sectorsWritten, err := strconv.ParseUint(fields[9], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse %q: %w", fields[3], err)
+		}
+
+		stats.WrittenBytes = sectorsWritten * 512
+
+		out[fields[2]] = stats
+	}
+
+	return out, nil
+}
+
+func getFilesystemMetrics(d *Daemon) (map[string]metrics.FilesystemMetrics, error) {
+	mounts, err := ioutil.ReadFile("/proc/mounts")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read /proc/mounts: %w", err)
+	}
+
+	out := map[string]metrics.FilesystemMetrics{}
+	scanner := bufio.NewScanner(bytes.NewReader(mounts))
+
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+
+		// Skip uninteresting mounts
+		if shared.StringInSlice(fields[2], defFSTypesExcluded) || defMountPointsExcluded.MatchString(fields[1]) {
+			continue
+		}
+
+		stats := metrics.FilesystemMetrics{}
+
+		stats.Mountpoint = fields[1]
+
+		statfs, err := filesystem.StatVFS(stats.Mountpoint)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to stat %s: %w", stats.Mountpoint, err)
+		}
+
+		fsType, err := filesystem.FSTypeToName(statfs.Type)
+		if err == nil {
+			stats.FSType = fsType
+		}
+
+		stats.SizeBytes = statfs.Blocks * uint64(statfs.Bsize)
+		stats.SizeBytes = statfs.Blocks * uint64(statfs.Bsize)
+		stats.SizeBytes = statfs.Blocks * uint64(statfs.Bsize)
+		stats.SizeBytes = statfs.Blocks * uint64(statfs.Bsize)
+
+		out[fields[0]] = stats
+	}
+
+	return out, nil
+}
+
+func getMemoryMetrics(d *Daemon) (metrics.MemoryMetrics, error) {
+	content, err := ioutil.ReadFile("/proc/meminfo")
+	if err != nil {
+		return metrics.MemoryMetrics{}, errors.Wrap(err, "Failed to read /proc/meminfo")
```

Comment:
Should use `fmt.Errorf` with the `%w` format instead of `errors.Wrap` as we're phasing it out.